src/pudl/helpers.py

```diff
@@ -339,6 +339,108 @@ def is_doi(doi):
+            ``right_date_col`` and ``on_cols`` argument.
+        left_date_col: Column in ``left`` containing datetime like data. Default is
+            ``report_date``. Must be convertible to a Datetime using
+            :func:`pandas.to_datetime`
+        right_date_col: Column in ``right`` containing datetime like data. Default is
+            ``report_date``. Must be convertible to a Datetime using
+            :func:`pandas.to_datetime`
+        on_cols: Columns to merge on in addition to newly created temporal columns. Typically
+            ID columns like ``plant_id_eia``, ``generator_id`` or ``boiler_id``.
+        gran: The desired granularity that the less granular dataframe should be
+            broadcast to. For now, this is the granularity of the more granular
+            dataframe.
+        report_time_in_period: When data is reported in the less granular dataframe.
+            When equal to "first" this means data in the right dataframe is reported
+            on the first date in the time period. E.g. annual dataframes have January
+            1st as the date. This is what happens by defualt if only a year or year-month
+            are provided to :func:`pandas.to_datetime` as strings.
+        merge_type: How the dataframes should be merged.
+
+
+    Returns:
+        Merged contents of left and right input dataframes.
+
+    Raises:
+        ValueError: if ``left_date_col`` or ``right_date_col`` columns are missing from their
+            respective input dataframes.
+        ValueError: if any of the labels referenced in ``on_cols`` are missing from either
+            the left or right dataframes.
+
+    """
+    # function breaks up a datetime column into separate columns
+    def assign_date_cols(df, date_col_name):
+        df[f"{date_col_name}"] = pd.to_datetime(df[f"{date_col_name}"])
+        df = df.assign(
+            year=lambda x: x[f"{date_col_name}"].dt.year,
+            quarter=lambda x: x[f"{date_col_name}"].dt.quarter,
+            month=lambda x: x[f"{date_col_name}"].dt.month,
+            day=lambda x: x[f"{date_col_name}"].dt.day,
+        )
+        return df
+
+    # add rows with an extra year for each unique group of rows, will be dropped
+    # this fixes an off by one issue with upsampling
+    # need to change offset interval based on gran argument
+    if report_time_in_period == "first":
+        extra_year_df = (
+            right.groupby(on_cols)[right_date_col].max() + pd.offsets.YearBegin()
+        )
+    else:
+        extra_year_df = (
+            right.groupby(on_cols)[right_date_col].min() - pd.offsets.YearEnd()
+        )
+    extra_year_df = extra_year_df.reset_index()
+    extra_year_df["is_extra_row"] = True
+    right = pd.concat([right, extra_year_df])
+    # upsample to higher granularity
+    if report_time_in_period == "first":
+        right = right.set_index(right_date_col).groupby(on_cols).resample(gran).ffill()
```

Comment:
I'm not entirely sure why this `resample` call is so slow. I guess upsampling just takes a while. There are a lot of limitations of `resample` which makes it necessary to call `set_index` and then `groupby`. 

[See this post](https://stackoverflow.com/questions/68604681/upscalling-multiindex-dataframe-in-pandas-1-3)

I think this `resample` needs to go